File: ru/codes/ruby/chapter_backtracking/preorder_traversal_i_compact.rb
=begin
File: preorder_traversal_i_compact.rb
Created Time: 2024-05-22
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

require_relative '../utils/tree_node'
require_relative '../utils/print_util'

# ## Предварительный обход: пример 1 ###
def pre_order(root)
  return unless root

  # Записать решение
  $res << root if root.val == 7

  pre_order(root.left)
  pre_order(root.right)
end

### Driver Code ###
if __FILE__ == $0
  root = arr_to_tree([1, 7, 3, 4, 5, 6, 7])
  puts "\nИнициализация двоичного дерева"
  print_tree(root)

  # Предварительный обход
  $res = []
  pre_order(root)

  puts "\nВсе узлы со значением 7"
  p $res.map { |node| node.val }
end

File: vi/codes/cpp/chapter_backtracking/preorder_traversal_i_compact.cpp
/**
 * File: preorder_traversal_i_compact.cpp
 * Created Time: 2023-04-16
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

vector<TreeNode *> res;

/* Preorder traversal: Example 1 */
void preOrder(TreeNode *root) {
    if (root == nullptr) {
        return;
    }
    if (root->val == 7) {
        // Record solution
        res.push_back(root);
    }
    preOrder(root->left);
    preOrder(root->right);
}

/* Driver Code */
int main() {
    TreeNode *root = vectorToTree(vector<int>{1, 7, 3, 4, 5, 6, 7});
    cout << "\nInitialize binary tree" << endl;
    printTree(root);

    // Preorder traversal
    preOrder(root);

    cout << "\nOutput all nodes with value 7" << endl;
    vector<int> vals;
    for (TreeNode *node : res) {
        vals.push_back(node->val);
    }
    printVector(vals);
}
